shared/src/main/scala-2.10/scalajson.ast/package.scala

```diff
@@ -0,0 +1,344 @@
+package scalajson
+
+package object ast {
+
+  /**
+    * A regex that will match any valid JSON number for unlimited
+    * precision
+    */
+  @inline protected[ast] val jNumberRegex: String =
+    """-?(?:0|[1-9]\d*)(?:\.\d+)?(?:[eE][+-]?\d+)?"""
+
+  /**
+    * Finds the hashcode for a numeric JSON string.
+    *
+    * @author Matthew de Detrich
+    * @param value
+    * @return
+    */
+  private[ast] def numericStringHashcode(value: String): Int = {
+    var result = 31
+    val length = value.length
+    var i = 0
+
+    if (value(0) == '-') {
+      // Found a negative, increment by one
+      result = result * 31 + '-': Int
+      i = 1
+    }
+
+    var char = value(i)
+
+    // From now on, we can just traverse all the chars
+
+    var negativeFlag = false
+
+    while (i < length) {
+      char = value(i)
+      // if char is e, lowercase it
+      if ((char | 0x20) == 'e') {
+
+        if (value(i + 1) == '-') {
+          // Found a negative, increment by one
+          i += 1
+          char = value(i)
+          negativeFlag = true
+        } else if (value(i + 1) == '+') {
+          // Found a positive, ignore
+          i += 1
+          char = value(i)
+        }
+
+        i += 1
+        char = value(i)
+
+        // Need to skip all leading zeroes, possible with e
+        while (char == '0' && i < length) {
+          i += 1
+          if (i != length)
+            char = value(i) // Fencepost, possible that this can be last character
+        }
+
+        if (i < length) {
+          if (negativeFlag) {
+            result = result * 31 + '-': Int
+          }
+
+          result = 31 * result + 'e': Int
+        }
+      } else if (char == '.') {
+
+        i += 1
+        char = value(i)
+
+        while (char == '0' && i < length) {
+          i += 1
+          if (i != length)
+            char = value(i) // Fencepost, possible that this can be last character
+        }
+
+        if (i < length) {
+          result = 31 * result + '.': Int // The decimal is not finishing with a 0
+        }
+      } else {
+        result = 31 * result + char: Int
+      }
+
+      i += 1
+    }
+    result
+  }
+
+  /** Tests two non-empty strings for equality, assuming both are decimal representations of numbers.
+    *
+    * Note: if the two strings are NOT decimal representations of numbers, the results of this method are undefined.
+    * (It is likely but not guaranteed that the method will return `false` even if the two strings are identical.)
+    *
+    * Many thanks for @Ichoran (https://github.com/Ichoran) for the implementation
+
+    * @author Rex Kerr
+    * @see https://github.com/Ichoran/kse/blob/master/src/main/scala/jsonal/Jast.scala#L683-L917
+    */
+  private[ast] def numericStringEquals(a: String, b: String): Boolean = {
```

Comment:
`numericStringEquals` is designed to test for actual number equality however there is clearly a bug in it (@Ichoran are you able to help here)?

This is something that people asked for, i.e. they expect that `JNumber` does proper number equality. So yes its true that this is a weird hybrid, but the reasoning for using `String` in the first place is because no unbounded real number exists in Scala (else we would be using that which would implement its own proper number equality)